sensor/common/clusterentities/store_ips.go

```diff
@@ -0,0 +1,217 @@
+	ipMap map[net.IPAddress]map[string]struct{}
+	// reverseIpMap maps deployment ids to sets of IP addresses associated with this deployment.
+	reverseIPMap map[string]map[net.IPAddress]struct{}
+	// historicalIPs is mimicking ipMap: IP Address -> deploymentID -> historyStatus
+	historicalIPs map[net.IPAddress]map[string]*entityStatus
+}
+
+func newIPsStoreWithMemory(numTicks uint16) *ipsStore {
+	store := &ipsStore{entitiesMemorySize: numTicks}
+	store.mutex.Lock()
+	defer store.mutex.Unlock()
+	store.initMapsNoLock()
+	return store
+}
+
+func (e *ipsStore) initMapsNoLock() {
+	e.ipMap = make(map[net.IPAddress]map[string]struct{})
+	e.reverseIPMap = make(map[string]map[net.IPAddress]struct{})
+	e.historicalIPs = make(map[net.IPAddress]map[string]*entityStatus)
+}
+
+func (e *ipsStore) resetMaps() {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	// Maps holding historical data must not be wiped on reset! Instead, all entities must be marked as historical.
+	// Must be called before the respective source maps are wiped!
+	// Performance optimization: no need to handle history if history is disabled
+	if e.entitiesMemorySize == 0 {
+		e.initMapsNoLock()
+		return
+	}
+	e.markAllIPsHistorical()
+
+	e.ipMap = make(map[net.IPAddress]map[string]struct{})
+	e.reverseIPMap = make(map[string]map[net.IPAddress]struct{})
+	e.updateMetricsNoLock()
+}
+
+func (e *ipsStore) updateMetricsNoLock() {
+}
+
+func (e *ipsStore) RecordTick() {
+	for ip, m := range e.historicalIPs {
+		for deploymentID, status := range m {
+			status.recordTick()
+			// Remove all historical entries that expired in this tick.
+			e.removeHistoricalExpiredIPs(deploymentID, ip)
+		}
+	}
+}
+
+func (e *ipsStore) Apply(
+	updates map[string]*EntityData,
+	incremental bool,
+	incPublicIPRefNoLock func(addr net.IPAddress),
+	decPublicIPRefNoLock func(addr net.IPAddress),
+) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	e.applyNoLock(updates, incremental, incPublicIPRefNoLock, decPublicIPRefNoLock)
+}
+
+func (e *ipsStore) applyNoLock(updates map[string]*EntityData, incremental bool,
+	incPublicIPRefNoLock func(addr net.IPAddress), decPublicIPRefNoLock func(addr net.IPAddress),
+) {
+	defer e.updateMetricsNoLock()
+	if !incremental {
+		for deploymentID := range updates {
+			e.purgeNoLock(deploymentID, decPublicIPRefNoLock)
+		}
+	}
+	for deploymentID, data := range updates {
+		if data == nil {
+			continue
+		}
+		e.applySingleNoLock(deploymentID, *data, incPublicIPRefNoLock)
+	}
+}
+
+func (e *ipsStore) purgeNoLock(deploymentID string, decPublicIPRefNoLock func(addr net.IPAddress)) {
+	for ip := range e.reverseIPMap[deploymentID] {
+		e.markHistoricalIP(deploymentID, ip)
+		// For entitiesMemorySize > 0, the deletion of historical expired entries happens after a tick.
+		// If memory is disabled, we should not wait for a tick and delete them immediately.
+		e.removeHistoricalExpiredIPs(deploymentID, ip)
+
+		delete(e.ipMap, ip)
```

Comment:
Similar to the endpoints store. This seems confusing. I think in this case it works because in line 105 we delete from the `ipMap` unconditionally in the endpoints store we only delete if `len(e.endpointMap[ep]) == 0`.